modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -4683,3 +4688,150 @@ def compare(self, other, **kwargs):
+                if copartition:
+                    _, list_of_right_parts, joined_index, row_lengths = (
+                        frame._copartition(
+                            Axis.ROW_WISE.value,
+                            data,
+                            how="left",
+                            sort=False,
+                            fill_value=True if i == 0 else None,
+                        )
+                    )
+                    data = frame.__constructor__(
+                        list_of_right_parts[0],
+                        joined_index,
+                        data.columns,
+                        row_lengths,
+                        data.column_widths,
+                    )
+                c[i] = data
+            new_caselist.append(tuple(c))
+
+        def map_data(
+            part_offset,
+            part_len,
+            data,
+            data_offset,
+            fill_value,
+            trivial=[is_trivial_idx],
+        ):
+            if isinstance(data, type(frame)):
+                if part_offset < len(data._partitions):
+                    return data._partitions[part_offset][0]._data
+                else:
+                    return [fill_value] * part_len
+
+            if isinstance(data, pandas.Series):
+                if trivial[0] is None:
+                    trivial[0] = is_trivial_index(frame.index)
+                if trivial[0] and is_trivial_index(data.index):
+                    data = data[data_offset : data_offset + part_len]
+                    diff = part_len - len(data)
+                    if diff > 0:
+                        data = pandas.concat((data, pandas.Series([fill_value] * diff)))
+                    return data
+                return data.reindex(
+                    frame.index[data_offset : data_offset + part_len],
+                    fill_value=fill_value,
+                )
+
+            if callable(data) or not is_list_like(data):
+                return data
+
+            return data[data_offset : data_offset + part_len]
+
+        new_parts = []
+        data_offset = 0
+        for i in range(0, parts_len):
+            part = parts[i]
+            part_len = lengths[i]
+            cases = [
+                tuple(
+                    map_data(i, part_len, d[0], data_offset, d[1])
+                    for d in zip(c, (True, None))
+                )
+                for c in new_caselist
+            ]
+            new_parts.append(
+                part.add_to_apply_calls(
+                    _case_when,
+                    cases,
+                    length=part_len,
+                    width=1,
+                )
+            )
+            data_offset += part_len
+        new_parts = np.array([[p] for p in new_parts])
+        frame = frame.__constructor__(
+            new_parts, columns=frame.columns, index=frame.index
+        )
+        return self.__constructor__(frame)
+
+
+def _case_when(df, caselist):  # pragma: no cover
```

Comment:
Usually we declare such wrappers inside concrete methods.

cc @dchigarev

Reply:
There is a function cache in RayWrapper.put(). Nested functions are not cached.